exporter/stefexporter/exporter.go

```diff
@@ -5,26 +5,395 @@ package stefexporter // import "github.com/open-telemetry/opentelemetry-collecto
+	s.lastAckedID = 0
+	s.interruptAckWaits = false
+	s.ackExpectedBy = s.ackExpectedBy[:0]
+
+	// Connect to the server.
+	var err error
+	s.grpcConn, err = s.cfg.ClientConfig.ToClientConn(ctx, host, s.set)
+	if err != nil {
+		return err
+	}
+
+	// Open a STEF/gRPC stream to the server.
+	grpcClient := stef_proto.NewSTEFDestinationClient(s.grpcConn)
+
+	// Let server know about our schema.
+	schema, err := oteltef.MetricsWireSchema()
+	if err != nil {
+		return err
+	}
+
+	settings := stefgrpc.ClientSettings{
+		Logger:       &loggerWrapper{s.logger},
+		GrpcClient:   grpcClient,
+		ClientSchema: schema,
+		Callbacks: stefgrpc.ClientCallbacks{
+			OnAck: s.onGrpcAck,
+		},
+	}
+	client := stefgrpc.NewClient(settings)
+
+	grpcWriter, opts, err := client.Connect(context.Background())
+	if err != nil {
+		s.grpcConn.Close()
+		return err
+	}
+
+	opts.Compression = s.compression
+
+	// Create record writer over gRPC stream.
+	s.stefWriter, err = oteltef.NewMetricsWriter(grpcWriter, opts)
+	if err != nil {
+		return err
+	}
+
+	// Start background timing tasks.
+	s.stopTicker = make(chan struct{})
+	go s.ticker()
+
+	s.isConnected = true
+
+	return nil
+}
+
+func (s *stefExporter) disconnect() {
+	s.connMutex.Lock()
+	defer s.connMutex.Unlock()
+
+	if !s.isConnected {
+		return
+	}
+
+	s.logger.Debug("Disconnecting...")
+
+	// Notify ticker() to stop.
+	close(s.stopTicker)
+
+	// Interrupt any pending waitForAck() call. They can't succeed anymore.
+	s.ackDataMutex.Lock()
+	s.interruptAckWaits = true
+	s.ackExpectedBy = s.ackExpectedBy[:0]
+	s.ackCond.Broadcast()
+	s.ackDataMutex.Unlock()
+
+	// If there is an existing connection close it.
+	if s.grpcConn != nil {
+		if err := s.grpcConn.Close(); err != nil {
+			s.logger.Error("failed to close grpc connection", zap.Error(err))
+		}
+	}
+
+	// Wait for all in progress wait in goroutines to get the notification that
+	// we are disconnecting and exit.
+	s.inProgressWaits.Wait()
+
+	s.isConnected = false
+}
+
 func (s *stefExporter) Shutdown(_ context.Context) error {
+	s.disconnect()
 	return nil
 }
 
-func (s *stefExporter) pushMetrics(_ context.Context, _ pmetric.Metrics) error {
+func (s *stefExporter) exportMetrics(ctx context.Context, md pmetric.Metrics) error {
+	// Keep count of how many concurrent writes are in progress.
+	s.writesInProgress.Add(1)
+	defer s.writesInProgress.Add(-1)
+
+	// Write the metrics to STEF stream.
+	expectedAckID, err := s.writeMetrics(ctx, md)
+	if err != nil {
+		return err
+	}
+
+	// Wait for acknowledgement from destination.
+	if err := s.waitForAck(expectedAckID); err != nil {
+		return err
+	}
+
 	return nil
 }
+
+func (s *stefExporter) writeMetrics(ctx context.Context, md pmetric.Metrics) (uint64, error) {
+	if err := s.ensureConnected(ctx, s.host); err != nil {
+		return 0, err
+	}
+
+	// stefWriter is not safe for concurrent writing, protect it.
+	s.stefWriterMutex.Lock()
+	defer s.stefWriterMutex.Unlock()
+
+	converter := stefpdatametrics.OtlpToTEFUnsorted{}
```

Comment:
typo in name? TEF -> STEF?

Reply:
Yes, needs to be updated in the dependency.

Reply:
Will update in the next release.